File: frontend/src/services/feedbackService.js
import { getStore, setStore, generateId, delay, getCurrentUserId, KEYS } from './mockData'

export const submitFeedback = async (data) => {
  await delay(300)
  const userId = getCurrentUserId()

  const feedback = {
    _id: generateId(),
    plan_id: data.plan_id,
    user_id: userId,
    rating: data.rating,
    thumbs: data.helpful === true ? 'up' : data.helpful === false ? 'down' : null,
    comment: data.comment || '',
    created_at: new Date().toISOString(),
  }

  const feedbacks = getStore(KEYS.FEEDBACKS) || []
  feedbacks.push(feedback)
  setStore(KEYS.FEEDBACKS, feedbacks)

  return { data: { feedback_id: feedback._id, message: 'Feedback submitted successfully' } }
}

export const getPlanFeedback = async (planId) => {
  await delay(200)
  const feedbacks = (getStore(KEYS.FEEDBACKS) || []).filter((f) => f.plan_id === planId)
  const ratings = feedbacks.filter((f) => f.rating != null)
  const avg = ratings.length > 0 ? ratings.reduce((sum, f) => sum + f.rating, 0) / ratings.length : 0

  return {
    data: {
      feedbacks,
      stats: { avg_rating: avg, count: feedbacks.length },
    },
  }
}

File: backend/app/services/feedback_service.py
from app.models.feedback import FeedbackModel
from app.utils.helpers import serialize_doc


class FeedbackService:

    @staticmethod
    def submit_feedback(plan_id, user_id, data):
        rating = data.get('rating')
        if rating is not None and (not isinstance(rating, (int, float)) or rating < 1 or rating > 5):
            return {'error': 'Rating must be between 1 and 5'}, 400

        feedback_id = FeedbackModel.create_feedback(
            plan_id=plan_id,
            user_id=user_id,
            rating=rating,
            thumbs=data.get('thumbs'),
            comment=data.get('comment', ''),
            useful_steps=data.get('useful_steps', []),
            problematic_steps=data.get('problematic_steps', []),
        )

        return {
            'message': 'Feedback submitted successfully',
            'feedback_id': feedback_id,
        }, 201

    @staticmethod
    def get_plan_feedback(plan_id):
        feedbacks = FeedbackModel.get_by_plan_id(plan_id)
        avg_rating = FeedbackModel.get_plan_avg_rating(plan_id)

        return {
            'feedbacks': [serialize_doc(f) for f in feedbacks],
            'stats': avg_rating,
        }, 200
